File: src/main/java/nz/co/searchwellington/controllers/models/IndexModelBuilder.java
package nz.co.searchwellington.controllers.models;

import java.util.Date;
import java.util.List;

import javax.servlet.http.HttpServletRequest;

import nz.co.searchwellington.controllers.LoggedInUserFilter;
import nz.co.searchwellington.controllers.RssUrlBuilder;
import nz.co.searchwellington.controllers.models.helpers.ArchiveLinksService;
import nz.co.searchwellington.model.User;
import nz.co.searchwellington.model.frontend.FrontendResource;
import nz.co.searchwellington.repositories.ContentRetrievalService;
import nz.co.searchwellington.urls.UrlBuilder;

import org.springframework.web.servlet.ModelAndView;

public class IndexModelBuilder extends AbstractModelBuilder implements ModelBuilder {
	
	private static final int MAX_OWNED_TO_SHOW_IN_RHS = 4;
	private static final int NUMBER_OF_COMMENTED_TO_SHOW = 2;
		
	private ContentRetrievalService contentRetrievalService;
	private RssUrlBuilder rssUrlBuilder;
	private LoggedInUserFilter loggedInUserFilter;
	private UrlBuilder urlBuilder;
	private ArchiveLinksService archiveLinksService;
		
	public IndexModelBuilder(ContentRetrievalService contentRetrievalService,
			RssUrlBuilder rssUrlBuilder, LoggedInUserFilter loggedInUserFilter,
			UrlBuilder urlBuilder, ArchiveLinksService archiveLinksService) {
		this.contentRetrievalService = contentRetrievalService;
		this.rssUrlBuilder = rssUrlBuilder;
		this.loggedInUserFilter = loggedInUserFilter;
		this.urlBuilder = urlBuilder;
		this.archiveLinksService = archiveLinksService;
	}
	
	@Override
	public boolean isValid(HttpServletRequest request) {
		return request.getPathInfo().matches("^/$") || request.getPathInfo().matches("^/json$");
	}
	
	@Override
	public ModelAndView populateContentModel(HttpServletRequest request) {
		if (isValid(request)) {
			ModelAndView mv = new ModelAndView();		
			final List<FrontendResource> latestNewsitems = contentRetrievalService.getLatestNewsitems(MAX_NEWSITEMS);                
			mv.addObject("main_content", latestNewsitems);
			
			Date monthOfLastItem = monthOfLastItem(latestNewsitems);
			if (monthOfLastItem != null) {
				mv.addObject("main_content_moreurl", urlBuilder.getArchiveLinkUrl(monthOfLastItem));
			}			
			setRss(mv, rssUrlBuilder.getBaseRssTitle(), rssUrlBuilder.getBaseRssUrl());
			return mv;
		}
		return null;
	}

	
	@Override
	public void populateExtraModelConent(HttpServletRequest request, boolean showBroken, ModelAndView mv) {		
		populateCommentedNewsitems(mv);
		populateSecondaryJustin(mv);
		populateGeocoded(mv);
		populateFeatured(mv);
		populateUserOwnedResources(mv, loggedInUserFilter.getLoggedInUser());		
		archiveLinksService.populateArchiveLinks(mv, contentRetrievalService.getArchiveMonths());
	}
		
	@Override
	public String getViewName(ModelAndView mv) {
		return "index";
	}
		
	private Date monthOfLastItem(List<FrontendResource> latestNewsitems) {
		if (latestNewsitems.size() > 0) {
			FrontendResource lastNewsitem = latestNewsitems.get(latestNewsitems.size() - 1);
			if (lastNewsitem.getDate() != null) {
				return lastNewsitem.getDate();
			}
		}
		return null;
	}
	
	private void populateUserOwnedResources(ModelAndView mv, User loggedInUser) {
		 if (loggedInUser != null) {
			 final int ownedCount = contentRetrievalService.getOwnedByCount(loggedInUser);
			 if (ownedCount > 0) {
				 mv.addObject("owned", contentRetrievalService.getOwnedBy(loggedInUser, MAX_OWNED_TO_SHOW_IN_RHS));
				 if (ownedCount > MAX_OWNED_TO_SHOW_IN_RHS) {
					 mv.addObject("owned_moreurl", urlBuilder.getProfileUrl(loggedInUser));
				 }
			 }			 
		 }
	}

	
	private void populateFeatured(ModelAndView mv) {
        mv.addObject("featured", contentRetrievalService.getFeaturedSites());
    }
	

	private void populateCommentedNewsitems(ModelAndView mv) {
		final List<FrontendResource> recentCommentedNewsitems = contentRetrievalService.getCommentedNewsitems(NUMBER_OF_COMMENTED_TO_SHOW + 1, 0);
		if (recentCommentedNewsitems.size() <= NUMBER_OF_COMMENTED_TO_SHOW) {
			mv.addObject("commented_newsitems", recentCommentedNewsitems);
		} else {
			mv.addObject("commented_newsitems", recentCommentedNewsitems.subList(0, NUMBER_OF_COMMENTED_TO_SHOW));
			mv.addObject("commented_newsitems_moreurl", "comment");
		}
	}
	
	
	private void populateGeocoded(ModelAndView mv) {
        List<FrontendResource> geocoded = contentRetrievalService.getGeocoded(0, MAX_NUMBER_OF_GEOTAGGED_TO_SHOW);
        if (geocoded.size() > 0) {
            mv.addObject("geocoded", geocoded);
        }
    }
	
	private void populateSecondaryJustin(ModelAndView mv) {
		mv.addObject("secondary_heading", "Just In");
		mv.addObject("secondary_description", "New additions.");
		mv.addObject("secondary_content", contentRetrievalService.getLatestWebsites(4));   
		mv.addObject("secondary_content_moreurl", "justin");        
	}
	
}

File: src/main/scala/nz/co/searchwellington/controllers/models/helpers/IndexModelBuilder.scala
package nz.co.searchwellington.controllers.models.helpers

import io.opentelemetry.api.trace.Span
import jakarta.servlet.http.HttpServletRequest
import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.filters.RequestPath
import nz.co.searchwellington.model.User
import nz.co.searchwellington.model.frontend.{Action, FrontendResource}
import nz.co.searchwellington.model.helpers.ArchiveLinksService
import nz.co.searchwellington.repositories.ContentRetrievalService
import nz.co.searchwellington.urls.{RssUrlBuilder, UrlBuilder}
import org.joda.time.DateTime
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import org.springframework.ui.ModelMap

import scala.collection.mutable.ListBuffer
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._

@Component class IndexModelBuilder @Autowired()(val contentRetrievalService: ContentRetrievalService, rssUrlBuilder: RssUrlBuilder,
                                                val urlBuilder: UrlBuilder, archiveLinksService: ArchiveLinksService,
                                                commonAttributesModelBuilder: CommonAttributesModelBuilder) extends ModelBuilder
  with CommonSizes with ReasonableWaits with ArchiveMonths {

  private val MAX_OWNED_TO_SHOW_IN_RHS = 4

  def isValid(request: HttpServletRequest): Boolean = {
    val path = RequestPath.getPathFrom(request)
    path.matches("^/$") ||
      path.matches("^/json$") ||
      path.matches("^/rss$")
  }

  def populateContentModel(request: HttpServletRequest, loggedInUser: Option[User])(implicit ec: ExecutionContext, currentSpan: Span): Future[Option[ModelMap]] = {
    val eventualLatestNewsitems = contentRetrievalService.getLatestNewsitems(MAX_NEWSITEMS, loggedInUser = loggedInUser)
    val eventualRecentlyAccepted = contentRetrievalService.getAcceptedNewsitems(MAX_NEWSITEMS, loggedInUser = loggedInUser, acceptedDate = None,
      acceptedAfter = Some(DateTime.now.minusDays(1)),
      publishedAfter = Some(DateTime.now.minusWeeks(2))
    ).map(a => (a._1.sortBy(_.date), a._2))
    for {
      latestNewsitems <- eventualLatestNewsitems
      recentlyAccepted <- eventualRecentlyAccepted
    } yield {
      val recentlyAcceptedToPromote = recentlyAccepted._1.filterNot(latestNewsitems.contains)
      val mainNewsitems = latestNewsitems ++ recentlyAcceptedToPromote

      val mv = new ModelMap().
        addAttribute("heading", "Wellynews").
        addAttribute("description", "Wellington related newsitems").
        addAttribute("link", urlBuilder.fullyQualified(urlBuilder.getHomeUri)).
        addAttribute(MAIN_CONTENT, mainNewsitems.asJava)

      monthOfLastItem(latestNewsitems).map { month =>
        mv.addAttribute("main_content_moreurl", urlBuilder.getIntervalUrl(month))
      }

      commonAttributesModelBuilder.setRss(mv, rssUrlBuilder.getBaseRssTitle, rssUrlBuilder.getBaseRssUrl)

      val submitActions = ListBuffer(
        Action("Submit website", urlBuilder.getSubmitWebsiteUrl),
        Action("Submit newsitem", urlBuilder.getSubmitNewsitemUrl),
        Action("Submit feed", urlBuilder.getSubmitFeedUrl)
      )
      if (loggedInUser.exists(_.isAdmin)) {
        submitActions.append(Action("Submit watchlist item", urlBuilder.getSubmitWatchlistUrl))
      }
      mv.addAttribute("submitActions", submitActions.asJava)

      Some(mv)
    }
  }

  def populateExtraModelContent(request: HttpServletRequest, loggedInUser: Option[User])(implicit ec: ExecutionContext, currentSpan: Span): Future[ModelMap] = {
    def populateUserOwnedResources(loggedInUser: Option[User]): Future[ModelMap] = {
      val mv = new ModelMap()
      loggedInUser.map { loggedInUser =>
        for {
          owned <- contentRetrievalService.getOwnedBy(loggedInUser, Some(loggedInUser), MAX_OWNED_TO_SHOW_IN_RHS)
        } yield {
          if (owned._2 > 0) {
            mv.addAttribute("owned", owned._1.asJava)
            if (owned._2 > MAX_OWNED_TO_SHOW_IN_RHS) {
              mv.addAttribute("owned_moreurl", urlBuilder.getProfileUrlFromProfileName(loggedInUser.getProfilename))
            }
          }
          mv
        }
      }.getOrElse {
        Future.successful(mv)
      }
    }

    val eventualWebsites = contentRetrievalService.getLatestWebsites(4, loggedInUser = loggedInUser)
    val eventualArchiveMonths = contentRetrievalService.getArchiveMonths(loggedInUser)
    val eventualArchiveStatistics = contentRetrievalService.getArchiveTypeCounts(loggedInUser)
    val eventualGeocodedNewsitems = contentRetrievalService.getGeocodedNewsitems(MAX_NUMBER_OF_GEOTAGGED_TO_SHOW, loggedInUser)
    val eventualUserOwnedResources = populateUserOwnedResources(loggedInUser)

    for {
      websites <- eventualWebsites
      archiveMonths <- eventualArchiveMonths
      archiveStatistics <- eventualArchiveStatistics
      geocodedNewsitems <- eventualGeocodedNewsitems
      userOwnedResources <- eventualUserOwnedResources
    } yield {
      new ModelMap()
        .addAllAttributes(secondaryJustin(websites._1))
        .addAllAttributes(populateGeocoded(geocodedNewsitems._1))
        .addAllAttributes(archiveLinksService.populateArchiveLinks(archiveMonths, archiveStatistics.toMap))
        .addAllAttributes(userOwnedResources)
    }
  }

  def getViewName(mv: ModelMap, loggedInUser: Option[User]): String = "index"

  private def populateGeocoded(geocoded: Seq[FrontendResource]): ModelMap = {
    if (geocoded.nonEmpty) {
      new ModelMap().addAttribute("geocoded", geocoded.asJava)
    } else {
      new ModelMap()
    }
  }

  private def secondaryJustin(websites: Seq[FrontendResource]): ModelMap = {
    new ModelMap().addAttribute("secondary_heading", "Just In")
      .addAttribute("secondary_description", "New additions.")
      .addAttribute("secondary_content", websites.asJava)
      .addAttribute("secondary_content_moreurl", "justin")
  }

}
